BrainPortal/cbrain_plugins/cbrain-plugins-base/userfiles/csv_file/views/_csv_file.html.erb

```diff
@@ -22,107 +22,122 @@
-        <%= hidden_field_tag :page, page %>
-        <%= hidden_field_tag :viewer, @viewer.name %>
-        <%= submit_tag "Refresh" %>
-      </span>
-    <% end %>
-
-    </br>
-
-    <div class="no_ajax_pagination">
-      <span>
-        <%= will_paginate csv_array,
-          :params => { :controller => :userfiles, :action => :show, :sort_index => @sort_index },
-          :container => false
-        %>
-        (<%= pluralize nb_row, "rows" %>)
-      </span>
-    </div>
+<div id="sub_viewer_csvfile">
+  <%= stylesheet_link_tag    'tablesorter_themes/blue/style' %>
+  <%= javascript_include_tag 'jquery.tablesorter.min' %>
+  <%= javascript_include_tag 'jquery.metadata' %>
+
+  <% if @userfile.is_locally_synced? %>
+    <% quote, separator, invalid_encoding = @userfile.guess_csv_quote_sep %>
 
-    <table id="report_table" class="tablesorter">
+    <% if ! (quote && separator) %>
+        Can't identify CSV quote character and separator character.
+    <% else %>
       <%
-        ################################
-        # Top headers of table
-        ################################
+        csv_array     = @userfile.create_csv_array(quote,separator)
+        with_header   = params[:with_header] == "1" ? true : false
+        per_page      = 500
+        nb_col        = csv_array[0].size
+        nb_row        = csv_array.size
+        do_sorting    = nb_row <= per_page
+        headers       = with_header ? csv_array.shift : do_sorting ? Array.new(nb_col, "---") : []
+        page          = (params[:page] || 1).to_i
+        page          = 1 if page < 1
+        csv_array     = WillPaginate::Collection.create(page, per_page) do |pager|
+          pager.replace(csv_array[(page-1)*per_page, per_page] || [])
+          pager.total_entries = csv_array.size
+          pager
+        end
       %>
 
-      <% if ! headers.empty? %>
-        <thead>
-          <tr>
-            <% headers.each do |col_name| %>
-              <th><%= col_name %></th>
-            <% end %>
-          </tr>
-        </thead>
+      <% if invalid_encoding %>
+        <span class="warning">
+          Warning: Invalid UTF-8 characters found. Some characters may not display correctly.
+        </span>
       <% end %>
 
       <%
-        ################################
-        # Main body of table
-        ################################
+        file_name     = params[:file_name] || @userfile.name
+        matched_class = SingleFile.descendants.unshift(SingleFile).find { |c| file_name =~ c.file_name_pattern }
+        viewer        = matched_class.class_viewers.first.partial rescue nil
       %>
+      <%= on_click_ajax_replace(
+          { :url     => display_userfile_path(@userfile,
+                                             :file_name             => file_name,
+                                             :action                => :display,
+                                             :viewer                => viewer,
```

Comment:
Although it is too late to change this, the code should really be clear about view names and viewer objects:

```ruby
viewer = (Viewer object)
viewer_name = viewer.name
```

In the query parameters we have often `viewer=xyz` even though what it means is really `viewer_name=xyz`

But the code should be updated to make sure that here that variables that contain just a name are called 'viewer_name'.